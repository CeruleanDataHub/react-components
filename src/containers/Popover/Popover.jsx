import PropTypes from "prop-types";
import React, { useRef, useState } from "react";
import { usePopper } from "react-popper-2";
import styled from "styled-components";

const PopperContainer = styled.div`
  box-shadow: 0 0 5px rgba(0, 0, 0, 0.3);
  border-radius: 5px;
  background-color: white;
  padding: 20px;
  text-align: center;

  #arrow {
    position: absolute;
    width: 10px;
    height: 10px;
    &:after {
      content: " ";
      background-color: white;
      box-shadow: -1px -1px 1px rgba(0, 0, 0, 0.1);
      position: absolute;
      top: -25px;
      left: 0;
      transform: rotate(45deg);
      width: 10px;
      height: 10px;
    }
  }

  &[data-popper-placement^="top"] > #arrow {
    bottom: -30px;
    :after {
      box-shadow: 1px 1px 1px rgba(0, 0, 0, 0.1);
    }
  }
`;

export const Popover = ({ isOpen, containerRef, children }) => {
  const [arrowRef, setArrowRef] = useState(null);
  const popoverRef = useRef(null);
  const { styles } = usePopper(containerRef.current, popoverRef.current, {
    modifiers: [
      {
        name: "arrow",
        options: {
          element: arrowRef
        }
      },
      {
        name: "offset",
        options: {
          offset: [0, 10]
        }
      }
    ]
  });

  if (!isOpen) {
    return null;
  }

  return (
    <PopperContainer ref={popoverRef} style={styles.popper}>
      <div ref={setArrowRef} style={styles.arrow} id="arrow" />
      {children}
    </PopperContainer>
  );
};

Popover.propTypes = {
  /** Defines whether to render the popover */
  isOpen: PropTypes.bool,
  /** Reference for container element  */
  containerRef: PropTypes.oneOfType([
    PropTypes.func,
    // eslint-disable-next-line no-undef
    PropTypes.shape({ current: PropTypes.instanceOf(Element) })
  ]),
  /** React node */
  children: PropTypes.node
};

Popover.defaultProps = {
  isOpen: false,
  containerRef: null,
  children: ""
};
