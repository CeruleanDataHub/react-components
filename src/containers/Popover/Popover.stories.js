import { withKnobs } from "@storybook/addon-knobs";
import React, { useRef, useState } from "react";
import styled from "styled-components";

import { Icon } from "../Icon/Icon";
import { Typography } from "../Typography/Typography";
import { Popover } from "./Popover";

export default {
  title: "Popover",
  component: Popover,
  decorators: [withKnobs]
};

const ListWrapper = styled.div`
  width: 100%;
  text-align: start;
`;

const PopoverContentList = styled.ul`
  list-style: none;
  margin: 0;
  padding: 0;
  display: flex;
  flex-direction: column;
  align-items: flex-start;
`;

const PopoverContentListItem = styled.li`
  padding: 0.5em 3em 0.5em 1.25em;
  display: block;

  &:hover {
    text-decoration: underline;
    cursor: pointer;
    background: papayawhip;
  }
`;

const ListItemIcon = styled.span`
  font-size: 0.75em;
  margin-right: 0.5rem;
`;

const ListDivider = styled.div`
  margin: 0.25em 0;
  border: 1px solid #ddd;
`;

export const PopoverStory = () => {
  const containerRef = useRef(null);
  const [popoverOpen, setPopoverOpen] = useState(false);

  return (
    <div style={{ height: 350 }}>
      <Typography fontFamily="openSans">
        <button
          type="button"
          ref={containerRef}
          onClick={() => setPopoverOpen(!popoverOpen)}
        >
          Toggle popover
        </button>
        <Popover isOpen={popoverOpen} containerRef={containerRef}>
          <PopoverContentList>
            <ListWrapper>
              <PopoverContentListItem>Profile</PopoverContentListItem>
              <PopoverContentListItem>Settings</PopoverContentListItem>
              <ListDivider />
              <PopoverContentListItem>
                <Icon name="rocket" as={ListItemIcon} />
                Manage devices
              </PopoverContentListItem>
              <PopoverContentListItem>
                <Icon name="user" as={ListItemIcon} />
                Manage users
              </PopoverContentListItem>
              <PopoverContentListItem>
                <Icon name="users" as={ListItemIcon} />
                Manage roles
              </PopoverContentListItem>
              <ListDivider />
              <PopoverContentListItem>Help</PopoverContentListItem>
              <PopoverContentListItem>Log out</PopoverContentListItem>
            </ListWrapper>
          </PopoverContentList>
        </Popover>
      </Typography>
    </div>
  );
};

export const PopoverOpenStory = () => {
  const containerRef = useRef(null);

  return (
    <div style={{ height: 100 }}>
      <button type="button" ref={containerRef}>
        Popover is open under the button
      </button>
      <Popover isOpen containerRef={containerRef}>
        Popover
      </Popover>
    </div>
  );
};
