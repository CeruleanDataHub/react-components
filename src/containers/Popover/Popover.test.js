import React, { useRef } from "react";
import renderer from "react-test-renderer";

import { Popover } from "./Popover";

// eslint-disable-next-line react/prop-types
const Wrapper = ({ isOpen }) => {
  const containerRef = useRef(null);
  const popoverRef = useRef(null);

  return (
    <>
      <button type="button" ref={containerRef}>
        Button
      </button>
      <Popover
        isOpen={isOpen}
        containerRef={containerRef}
        popperRef={popoverRef}
      >
        <div>Popover test content</div>
      </Popover>
    </>
  );
};

describe("Popover", () => {
  it("should render closed", () => {
    const component = renderer.create(<Wrapper />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should render open", () => {
    const component = renderer.create(<Wrapper isOpen />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
