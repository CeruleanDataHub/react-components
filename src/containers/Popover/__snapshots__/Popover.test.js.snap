// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover should render closed 1`] = `
<button
  type="button"
>
  Button
</button>
`;

exports[`Popover should render open 1`] = `
Array [
  <button
    type="button"
  >
    Button
  </button>,
  <div
    className="sc-AxjAm fOWUxG"
    style={
      Object {
        "left": "0",
        "position": "absolute",
        "top": "0",
      }
    }
  >
    <div
      id="arrow"
    />
    <div>
      Popover test content
    </div>
  </div>,
]
`;
